master/internal/task/allocation_service.go

```diff
@@ -0,0 +1,306 @@
+
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/internal/api"
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/rm"
+	"github.com/determined-ai/determined/master/internal/sproto"
+	"github.com/determined-ai/determined/master/internal/task/allgather"
+	"github.com/determined-ai/determined/master/internal/task/preemptible"
+	"github.com/determined-ai/determined/master/pkg/actor"
+	detLogger "github.com/determined-ai/determined/master/pkg/logger"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/tasks"
+)
+
+var syslog = logrus.WithField("component", "allocation_service")
+
+// DefaultService is the singleton default allocationService.
+var DefaultService AllocationService = newAllocationService()
+
+// allocationService is used to launch, track and interact with allocations.
+type allocationService struct {
+	syslog      *logrus.Entry
+	mu          sync.RWMutex
+	allocations map[model.AllocationID]*allocation
+}
+
+// newAllocationService creates a new allocationService.
+func newAllocationService() *allocationService {
+	return &allocationService{
+		syslog:      logrus.WithField("component", "allocation-service"),
+		allocations: map[model.AllocationID]*allocation{},
+	}
+}
+
+// StartAllocation starts an allocation and returns a handle to it.
+func (as *allocationService) StartAllocation(
+	logCtx detLogger.Context,
+	req sproto.AllocateRequest,
+	db db.DB,
+	rm rm.ResourceManager,
+	specifier tasks.TaskSpecifier,
+	system *actor.System,
+	onExit func(*AllocationExited),
+) error {
+	as.mu.Lock()
+	defer as.mu.Unlock()
+
+	ref, err := newAllocation(logCtx, req, db, rm, specifier, system)
+	if err != nil {
+		return err
+	}
+	as.allocations[req.AllocationID] = ref
+
+	go func() {
+		exit := ref.awaitTermination()
+		if err := ref.Close(); err != nil {
+			syslog.WithError(err).Error("cleaning up allocation")
+		}
+
+		as.mu.Lock()
+		defer as.mu.Unlock()
+		delete(as.allocations, req.AllocationID)
+		onExit(exit)
+	}()
+	return nil
+}
+
+// GetAllAllocationIDs returns all registered allocation ids.
+func (as *allocationService) GetAllAllocationIDs() []model.AllocationID {
+	as.mu.RLock()
+	defer as.mu.RUnlock()
+	return maps.Keys(as.allocations)
+}
+
+// SendLog sends a container log, enriched with metadata from the allocation.
+func (as *allocationService) SendLog(
+	ctx context.Context,
+	id model.AllocationID,
+	log *sproto.ContainerLog,
+) {
+	ref, err := as.getAllocation(id)
+	if err != nil {
+		syslog.Warnf("dropped log for unknown allocation: %s", err)
+		return
+	}
+	ref.SendContainerLog(log)
+}
+
+// SetReady sets the ready bit and moves the allocation to the running state if it has not
+// progressed past it already.
+func (as *allocationService) SetReady(ctx context.Context, id model.AllocationID) error {
+	ref, err := as.waitForRestore(ctx, id)
```

Comment:
This isn't saying this is a problem just curious what we would need to do at this point to remove these ```waitForRestore```?

Reply:
uhh... i'm not sure. maybe once restore is instant. or it is gated globally and we can return 503 in middleware until restore is done, and make everything retry.